File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/browsing/items/list/ItemListFragment.java
package com.lasthopesoftware.bluewater.client.browsing.items.list;

import android.app.Activity;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewGroup.LayoutParams;
import android.widget.ListView;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;

import androidx.fragment.app.Fragment;

import com.lasthopesoftware.bluewater.client.browsing.items.IItem;
import com.lasthopesoftware.bluewater.client.browsing.items.Item;
import com.lasthopesoftware.bluewater.client.browsing.items.access.ItemProvider;
import com.lasthopesoftware.bluewater.client.browsing.items.list.menus.changes.handlers.IItemListMenuChangeHandler;
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.access.parameters.FileListParameters;
import com.lasthopesoftware.bluewater.client.browsing.library.access.ILibraryProvider;
import com.lasthopesoftware.bluewater.client.browsing.library.access.LibraryRepository;
import com.lasthopesoftware.bluewater.client.browsing.library.access.session.ProvideSelectedLibraryId;
import com.lasthopesoftware.bluewater.client.browsing.library.access.session.SelectedBrowserLibraryIdentifierProvider;
import com.lasthopesoftware.bluewater.client.browsing.library.views.handlers.OnGetLibraryViewItemResultsComplete;
import com.lasthopesoftware.bluewater.client.connection.HandleViewIoException;
import com.lasthopesoftware.bluewater.client.connection.selected.SelectedConnection;
import com.lasthopesoftware.bluewater.client.stored.library.items.StoredItemAccess;
import com.lasthopesoftware.bluewater.shared.exceptions.UnexpectedExceptionToasterResponse;
import com.lasthopesoftware.bluewater.shared.promises.extensions.LoopedInPromise;
import com.namehillsoftware.handoff.promises.response.PromisedResponse;
import com.namehillsoftware.handoff.promises.response.VoidResponse;
import com.namehillsoftware.lazyj.AbstractSynchronousLazy;
import com.namehillsoftware.lazyj.CreateAndHold;

import org.jetbrains.annotations.NotNull;

import java.util.List;

public class ItemListFragment extends Fragment {

	private static final String ARG_CATEGORY_POSITION = "category_position";

	private IItemListMenuChangeHandler itemListMenuChangeHandler;

	private final CreateAndHold<ListView> lazyListView = new AbstractSynchronousLazy<ListView>() {
		@Override
		protected ListView create() {
			final ListView listView = new ListView(getActivity());
			listView.setVisibility(View.INVISIBLE);
			return listView;
		}
	};

	private final CreateAndHold<ProgressBar> lazyProgressBar = new AbstractSynchronousLazy<ProgressBar>() {
		@Override
		protected ProgressBar create() {
			final ProgressBar pbLoading = new ProgressBar(getActivity(), null, android.R.attr.progressBarStyleLarge);
			final RelativeLayout.LayoutParams pbParams = new RelativeLayout.LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT);
			pbParams.addRule(RelativeLayout.CENTER_IN_PARENT);
			pbLoading.setLayoutParams(pbParams);
			return pbLoading;
		}
	};

	private final CreateAndHold<RelativeLayout> lazyLayout = new AbstractSynchronousLazy<RelativeLayout>() {
		@Override
		protected RelativeLayout create() {
			final Activity activity = getActivity();

			final RelativeLayout layout = new RelativeLayout(activity);
			layout.setLayoutParams(new RelativeLayout.LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT));

			layout.addView(lazyProgressBar.getObject());
			layout.addView(lazyListView.getObject());

			return layout;
		}
	};

	public static ItemListFragment getPreparedFragment(final int libraryViewId) {
		final ItemListFragment returnFragment = new ItemListFragment();
		final Bundle args = new Bundle();
		args.putInt(ItemListFragment.ARG_CATEGORY_POSITION, libraryViewId);
		returnFragment.setArguments(args);
		return returnFragment;
	}

	@Override
	public View onCreateView(@NotNull LayoutInflater inflater, final ViewGroup container, Bundle savedInstanceState) {
		return lazyLayout.getObject();
	}

	@Override
	public void onStart() {
		super.onStart();

		final Activity activity = getActivity();
		if (activity == null) return;

		lazyListView.getObject().setVisibility(View.INVISIBLE);
		lazyProgressBar.getObject().setVisibility(View.VISIBLE);

		final ILibraryProvider libraryProvider = new LibraryRepository(activity);
		final ProvideSelectedLibraryId selectedLibraryIdentifierProvider = new SelectedBrowserLibraryIdentifierProvider(activity);

		libraryProvider
			.getLibrary(selectedLibraryIdentifierProvider.getSelectedLibraryId())
			.then(new VoidResponse<>(activeLibrary -> {
				final PromisedResponse<List<Item>, Void> onGetVisibleViewsCompleteListener = LoopedInPromise.response(new VoidResponse<>(result -> {
					if (result == null || result.size() == 0) return;

					final int categoryPosition = getArguments().getInt(ARG_CATEGORY_POSITION);
					final IItem category = categoryPosition < result.size() ? result.get(categoryPosition) : result.get(result.size() - 1);

					fillStandardItemView(category);
				}), activity);

				final Runnable fillItemsRunnable = new Runnable() {

					@Override
					public void run() {
						SelectedConnection.getInstance(activity).promiseSessionConnection()
							.eventually(c -> ItemProvider.provide(c, activeLibrary.getSelectedView()))
							.eventually(onGetVisibleViewsCompleteListener)
							.excuse(new HandleViewIoException(activity, this))
							.eventuallyExcuse(LoopedInPromise.response(new UnexpectedExceptionToasterResponse(activity), activity));
					}
				};

				fillItemsRunnable.run();
			}));
	}

	private void fillStandardItemView(final IItem category) {
		final Activity activity = getActivity();
		if (activity == null) return;

		final ProvideSelectedLibraryId selectedLibraryIdentifierProvider = new SelectedBrowserLibraryIdentifierProvider(getContext());
		final ILibraryProvider libraryProvider = new LibraryRepository(getContext());

		libraryProvider
			.getLibrary(selectedLibraryIdentifierProvider.getSelectedLibraryId())
			.then(new VoidResponse<>(library -> {
				PromisedResponse<List<Item>, Void> onGetLibraryViewItemResultsComplete = LoopedInPromise.response(new OnGetLibraryViewItemResultsComplete(
					activity,
					lazyListView.getObject(),
					lazyProgressBar.getObject(),
					itemListMenuChangeHandler,
					FileListParameters.getInstance(),
					new StoredItemAccess(activity),
					library), activity);

				final Runnable fillItemsRunnable = new Runnable() {

					@Override
					public void run() {
						SelectedConnection.getInstance(activity).promiseSessionConnection()
							.eventually(c -> ItemProvider.provide(c, category.getKey()))
							.eventually(onGetLibraryViewItemResultsComplete)
							.excuse(new HandleViewIoException(activity, this))
							.eventuallyExcuse(LoopedInPromise.response(new UnexpectedExceptionToasterResponse(activity), activity));
					}
				};

				fillItemsRunnable.run();
			}));
	}

	public void setOnItemListMenuChangeHandler(IItemListMenuChangeHandler itemListMenuChangeHandler) {
		this.itemListMenuChangeHandler = itemListMenuChangeHandler;
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/browsing/items/list/ItemListFragment.kt
package com.lasthopesoftware.bluewater.client.browsing.items.list

import android.os.Bundle
import android.os.Handler
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import android.widget.RelativeLayout
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.lasthopesoftware.bluewater.R
import com.lasthopesoftware.bluewater.client.browsing.items.IItem
import com.lasthopesoftware.bluewater.client.browsing.items.Item
import com.lasthopesoftware.bluewater.client.browsing.items.ItemId
import com.lasthopesoftware.bluewater.client.browsing.items.access.CachedItemProvider
import com.lasthopesoftware.bluewater.client.browsing.items.list.menus.changes.handlers.IItemListMenuChangeHandler
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.access.parameters.FileListParameters
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.access.stringlist.ItemStringListProvider
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.access.stringlist.LibraryFileStringListProvider
import com.lasthopesoftware.bluewater.client.browsing.items.menu.ActivityLaunching
import com.lasthopesoftware.bluewater.client.browsing.library.access.LibraryRepository
import com.lasthopesoftware.bluewater.client.browsing.library.access.session.SelectedBrowserLibraryIdentifierProvider
import com.lasthopesoftware.bluewater.client.browsing.library.access.session.SelectedBrowserLibraryProvider
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.connection.HandleViewIoException
import com.lasthopesoftware.bluewater.client.connection.session.ConnectionSessionManager
import com.lasthopesoftware.bluewater.client.stored.library.items.StateChangeBroadcastingStoredItemAccess
import com.lasthopesoftware.bluewater.client.stored.library.items.StoredItemAccess
import com.lasthopesoftware.bluewater.settings.repository.access.CachingApplicationSettingsRepository.Companion.getApplicationSettingsRepository
import com.lasthopesoftware.bluewater.shared.android.view.ViewUtils
import com.lasthopesoftware.bluewater.shared.exceptions.UnexpectedExceptionToasterResponse
import com.lasthopesoftware.bluewater.shared.messages.application.ApplicationMessageBus.Companion.getApplicationMessageBus
import com.lasthopesoftware.bluewater.shared.messages.application.ScopedApplicationMessageBus
import com.lasthopesoftware.bluewater.shared.messages.registerReceiver
import com.lasthopesoftware.bluewater.shared.promises.extensions.LoopedInPromise.Companion.response
import com.lasthopesoftware.bluewater.shared.promises.extensions.keepPromise
import com.lasthopesoftware.bluewater.tutorials.TutorialManager

class ItemListFragment : Fragment() {
	private var itemListMenuChangeHandler: IItemListMenuChangeHandler? = null

	private val handler by lazy { Handler(requireContext().mainLooper) }

	private val browserLibraryIdProvider by lazy {
		SelectedBrowserLibraryIdentifierProvider(requireContext().getApplicationSettingsRepository())
	}

	private val selectedLibraryProvider by lazy {
		SelectedBrowserLibraryProvider(
			browserLibraryIdProvider,
			LibraryRepository(requireContext())
		)
	}

	private val itemListProvider by lazy {
		val connectionProvider = ConnectionSessionManager.get(requireContext())

		ItemStringListProvider(
            FileListParameters,
            LibraryFileStringListProvider(connectionProvider)
        )
	}

	private val itemProvider by lazy { CachedItemProvider.getInstance(requireContext()) }

	private val tutorialManager by lazy { TutorialManager(requireContext()) }

	private val applicationMessageBus = lazy {
		val applicationMessageBus = requireContext().getApplicationMessageBus()
		ScopedApplicationMessageBus(applicationMessageBus, applicationMessageBus).apply {
			registerReceiver { l: ActivityLaunching ->
				val isLaunching = l != ActivityLaunching.HALTED // Only show the item list view again when launching error'ed for some reason

				recyclerView?.visibility = ViewUtils.getVisibility(!isLaunching)
				progressBar?.visibility = ViewUtils.getVisibility(isLaunching)
			}
		}
	}

	private val demoableItemListAdapter by lazy {
		browserLibraryIdProvider.selectedLibraryId.then {
			it?.let { libraryId ->
				itemListMenuChangeHandler?.let { itemListMenuChangeHandler ->
					activity
						?.let { fa ->
							DemoableItemListAdapter(
								fa,
								applicationMessageBus.value,
								itemListProvider,
								itemListMenuChangeHandler,
								StoredItemAccess(fa),
								itemProvider,
								libraryId,
								tutorialManager
							)
						}
						?: requireContext()
							.let { context ->
								ItemListAdapter(
									context,
									applicationMessageBus.value,
									itemListProvider,
									itemListMenuChangeHandler,
									StateChangeBroadcastingStoredItemAccess(StoredItemAccess(context), applicationMessageBus.value),
									itemProvider,
									libraryId,
								)
							}
				}
			}
		}
	}

	private var recyclerView: RecyclerView? = null
	private var progressBar: ProgressBar? = null
	private var layout: RelativeLayout? = null

	override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
		layout = inflater.inflate(R.layout.asynchronous_recycler_view, container, false) as RelativeLayout
		progressBar = layout?.findViewById(R.id.items_loading_progress)
		recyclerView = layout?.findViewById(R.id.loaded_recycler_view)
		return layout
	}

	override fun onStart() {
		super.onStart()

		val context = requireContext()

		recyclerView?.visibility = ViewUtils.getVisibility(false)
		progressBar?.visibility = ViewUtils.getVisibility(true)

		selectedLibraryProvider
			.browserLibrary
			.then { activeLibrary ->
				if (activeLibrary == null) return@then

				val onGetVisibleViewsCompleteListener =
					response<List<Item>, Unit>({ result ->
						result
							.takeUnless { result.isEmpty() }
							?.let {
								arguments
									?.getInt(ARG_CATEGORY_POSITION)
									?.let { categoryPosition ->
										if (categoryPosition < result.size) result[categoryPosition]
										else result[result.size - 1]
									}
							}
							?.also { fillStandardItemView(activeLibrary.libraryId, it) }
					}, context)

				object : Runnable {
					override fun run() {
						itemProvider.promiseItems(activeLibrary.libraryId, ItemId(activeLibrary.selectedView))
							.eventually(onGetVisibleViewsCompleteListener)
							.excuse(HandleViewIoException(context, this))
							.eventuallyExcuse(response(UnexpectedExceptionToasterResponse(context), handler))
					}
				}.run()
			}
	}

	override fun onDestroy() {
		super.onDestroy()

		if (applicationMessageBus.isInitialized())
			applicationMessageBus.value.close()
	}

	private fun fillStandardItemView(libraryId: LibraryId, category: IItem) {
		demoableItemListAdapter
			.eventually(response({ adapter ->
				recyclerView
					?.takeIf { it.adapter == null || it.adapter != adapter }
					?.also {
						it.adapter = adapter
						val layoutManager = LinearLayoutManager(context)
						it.layoutManager = layoutManager
						it.addItemDecoration(DividerItemDecoration(context, layoutManager.orientation))
					}

				adapter?.also { ItemHydration(libraryId, category, adapter) }
			}, handler))
	}

	fun setOnItemListMenuChangeHandler(itemListMenuChangeHandler: IItemListMenuChangeHandler?) {
		this.itemListMenuChangeHandler = itemListMenuChangeHandler
	}

	companion object {
		private const val ARG_CATEGORY_POSITION = "category_position"

		@JvmStatic
		fun getPreparedFragment(libraryViewId: Int): ItemListFragment {
			val returnFragment = ItemListFragment()
			val args = Bundle()
			args.putInt(ARG_CATEGORY_POSITION, libraryViewId)
			returnFragment.arguments = args
			return returnFragment
		}
	}

	private inner class ItemHydration(private val libraryId: LibraryId, private val category: IItem, private val adapter: ItemListAdapter) : Runnable {
		init {
			run()
		}

		override fun run() {
			val context = requireContext()
			itemProvider.promiseItems(libraryId, ItemId(category.key))
				.eventually { i -> i?.let(adapter::updateListEventually).keepPromise(Unit) }
				.eventually(response({
					progressBar?.visibility = ViewUtils.getVisibility(false)
					recyclerView?.visibility = ViewUtils.getVisibility(true)
				}, handler))
				.excuse(HandleViewIoException(context, this))
				.eventuallyExcuse(response(UnexpectedExceptionToasterResponse(context), handler))
		}
	}
}
